File: Two-Pointers/2Sum.go
// Source: https://leetcode.com/problems/two-sum/
package Two_Pointers

import "sort"

func twoSum(nums []int, target int) []int {
	length := len(nums)
	for i := 0; i < length; i++ {
		for j := i + 1; j < length; j++ {
			if nums[i]+nums[j] == target {
				return []int{i, j}
			}
		}
	}
	return []int{}
}

// Time Complexity - O(n^2)

func twoSumOpt(nums []int, target int) []int {
	numsWithIndex := make([][2]int, len(nums))
	for i, num := range nums {
		numsWithIndex[i] = [2]int{num, i}
	}

	sort.Slice(numsWithIndex, func(i, j int) bool {
		return numsWithIndex[i][0] < numsWithIndex[j][0]
	})

	left, right := 0, len(nums)-1
	for left < right {
		sum := numsWithIndex[left][0] + numsWithIndex[right][0]
		if sum == target {
			return []int{numsWithIndex[left][1], numsWithIndex[right][1]}
		} else if sum < target {
			left++
		} else {
			right--
		}
	}

	return []int{}
}

// Time Complexity - O(nlogn)

File: Two-Pointers/2Sum.js
// LeetCode: https://leetcode.com/problems/two-sum/
// Example: 2Sum([2, 7, 11, 15], 9) => [0, 1]
let twoSum = function(nums, target) {
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i] + nums[j] === target) {
        return [i, j]
      }
    }
  }
};
// Time Complexity -  O(n^2)
